client/src/composables/historyDragDrop.ts

```diff
@@ -0,0 +1,173 @@
+            (destinationHistoryId.value && fromHistoryId.value === destinationHistoryId.value) ||
+            (!createNew && !destinationHistoryId.value) ||
+            !getDragItems().length ||
+            processingDrop.value
+    );
+
+    function getDragItems() {
+        const storeItems = eventStore.getDragItems();
+        // Filter out any non-history or `DatasetCollectionElement` items
+        return storeItems?.filter(
+            (item) => isHistoryItem(item) || isDatasetElement(item as DCESummary)
+        ) as DraggableHistoryItem[];
+    }
+
+    function onDragEnter(e: DragEvent) {
+        if (operationDisabled.value) {
+            return;
+        }
+        dragTarget.value = e.target;
+        showDropZone.value = true;
+    }
+
+    function onDragOver(e: DragEvent) {
+        if (operationDisabled.value) {
+            return;
+        }
+        e.preventDefault();
+    }
+
+    function onDragLeave(e: DragEvent) {
+        if (operationDisabled.value) {
+            return;
+        }
+        if (dragTarget.value === e.target) {
+            showDropZone.value = false;
+        }
+    }
+
+    async function onDrop() {
+        showDropZone.value = false;
+        if (operationDisabled.value) {
+            return;
+        }
+        processingDrop.value = true;
+
+        let datasetCount = 0;
+        let collectionCount = 0;
+
+        try {
+            const dragItems = getDragItems();
+            const multiple = dragItems.length > 1;
+            const originalHistoryId = fromHistoryId.value as string;
+
+            let historyId;
+            if (destinationHistoryId.value) {
+                historyId = destinationHistoryId.value;
+            } else if (createNew) {
+                await historyStore.createNewHistory();
+                historyId = historyStore.currentHistoryId;
+            } else {
+                historyId = null;
+            }
+            if (!historyId) {
+                throw new Error("Destination history not found or created");
+            }
+
+            // iterate over the data array and copy each item to the current history
+            for (const item of dragItems) {
+                let dataSource: HistoryContentSource;
+                let type: HistoryContentType;
+                let id: string;
+                if (isHistoryItem(item)) {
+                    dataSource = item.history_content_type === "dataset" ? "hda" : "hdca";
+                    type = item.history_content_type;
+                    id = item.id;
+                }
+                // For DCEs, only `DCEDataset`s are droppable, `DCEDatasetCollection`s are not
+                else if (isDatasetElement(item) && item.object) {
+                    dataSource = "hda";
+                    type = "dataset";
+                    id = item.object.id;
+                } else {
+                    throw new Error(`Invalid item type${item.element_type ? `: ${item.element_type}` : ""}`);
```

Comment:
There was an error handler in https://github.com/galaxyproject/galaxy/pull/18699/files#diff-68e87a439364cbb0acfc54a106c0830012c8005d4288102ba12a57fc7ffcedd6L408, you might want to re-instate that. This btw can happen if the collection isn't populated yet, so we'll definitely run into this from time to time.

Reply:
I see you added a handler in https://github.com/galaxyproject/galaxy/pull/18699/files#diff-ad4aa7272cdb314a83cba0349216030549831df46d0113ef0d5b60a67d72a26aR160 ... I think it might be better to do error handling like this in the caller, so context-specific actions can be taken (e.g. Toast vs inline errors in this example). Also the caller might want to know if the action succeeded, which is will always if you catch the error.